### DISCUSSION TITLE
Minimum bounding sphere computation for meshes/meshlets

### DISCUSSION THREAD

**Discussion by expenses** [zeux/meshoptimizer#636] [Ideas]
Hi, I've been using your library to compute meshlets and their bounds and it's been working out well so far. The current code for this uses a linear scan: https://github.com/zeux/meshoptimizer/blob/28bd987efa521b9d7301b91448fd32600067c2e9/src/clusterizer.cpp#L78-L146. This is good enough for just-in-time uses, but for offline computation it'd be great to use an implementation of [Welzl's algorithm](https://en.wikipedia.org/wiki/Smallest-circle_problem#Welzl's_algorithm) that calculates the exact minimum bounding sphere.

Something like gltfpack could compute and store the minimum bounding-sphere for meshes as well.

**Comment by zeux** [comment]
I'd need to go back to look at my notes; my recollection is that I considered Welzl and rejected it because of stability and performance/determinism concerns (it's amortized linear in theory but theory and practice tend to diverge); note that the currently used algorithm doesn't guarantee minimal results but it provides, generally, results that are fairly good. I think I experimented with other guaranteed-linear-time algorithms and this was the best among the ones that are based on heuristics.

Out of curiosity, do you have specific examples where the spheres are significantly suboptimal? I'm very open to improving this but I'd prefer to explore options that aren't Welzl.

**Reply by castano** [reply]
There's also Ritter's and EPOS algorithms: https://ep.liu.se/ecp/034/009/ecp083409.pdf
Here's an implementation of all of them: https://gist.github.com/castano/893695b3291d949242f2aee3943dfd55
Note there are some robustness issues in the Sphere constructor when the points are coplanar. I've been meaning to revisit and robustify that code, but haven't had a good excuse to do so.

**Reply by zeux** [reply]
Yeah the clusterizer code uses Ritter right now.

**Reply by expenses** [reply]
>Out of curiosity, do you have specific examples where the spheres are significantly suboptimal?

No, I'm just a little wary that some performance is left on the table. @castano thanks for much for posting these! I was having trouble finding drop-in implementations. I'll do some testing with my own meshes and report back.

**Reply by trifonov-borislav** [reply]
In some cases I was getting radius up to 20% larger than the optimal using Ritter, but the average case is not much worse. I implemented EPOS-6 before and with and works well (within a couple %) while it's still fast enough, but it relies on Welzl as a subroutine: basically you run Welzl on the projected by EPOS point subset. The trick is that Welzl has numerical issues and can explode, so you need to run Ritter beforehand and pass that in as an upper bound to detect such a situation (simply checking determinant is **not** sufficient). Definitely adds complexity but it's not too bad

**Comment by zeux** [comment]
FWIW I continue to have reservations about using Welzl but I've adjusted the code to use the axes EPOS uses (EPOS-14 variant to be more precise; I've tested EPOS-26 axis selection and the extra axes introduced there don't seem to meaningfully improve the radius on clusters I've tested) in https://github.com/zeux/meshoptimizer/pull/883.

On some meshes this results in 1-2% tighter average cluster bounds - I haven't measured the impact on worst case cluster bounds but I'd expect it to be more significant obviously. As a result this improves the cluster occlusion culling efficiency a little bit; the time to compute the bounds is a little longer but it's mitigated by small code tweaks and only using this for positions, as I didn't find the effect on cone angles to be significant.
